### PR TITLE
Introduce more strict compiler warnings and fix them

### PR DESCRIPTION
**Pull request by Ryex**
Add some reasonable compiler warnings to help prevent future problems.
clean up after them.

### CODE REVIEW

**Review by Scrumplex — APPROVED** [review]

**Comment by Scrumplex** [comment]
This might cause issues with Qt 5.12 < x < 5.15 as we only support these versions specifically. Some options might have been deprecated between 5.12 and 5.15 but only replaced for >=5.15 in our codebase

**Comment by Ryex** [comment]
> This might cause issues with Qt 5.12 < x < 5.15 as we only support these versions specifically. Some options might have been deprecated between 5.12 and 5.15 but only replaced for >=5.15 in our codebase

I'm not sure how this would break anything. none of these warning warn about deprecation. and they are only warnings. if they flag something that would fall in that feature gap we know to put it in pre-processor flags or special case it. right?

**Review by Trial97 — APPROVED** [review]

**Comment by Scrumplex** [comment]
> I'm not sure how this would break anything. none of these warning warn about deprecation

Let's say something was deprecated in Qt 5.14 in favor of something else and our code only handles this in the form of a `#if QT_VERSION >= QT_VERSION_CHECK(5, 15, 0)` check, building Prism Launcher against Qt 5.14 would raise a deprecation warning, which would then abort the build as you have added `-Werror` (and MSVC equivalents)

**Comment by Ryex** [comment]
> > I'm not sure how this would break anything. none of these warning warn about deprecation
> 
> Let's say something was deprecated in Qt 5.14 in favor of something else and our code only handles this in the form of a `#if QT_VERSION >= QT_VERSION_CHECK(5, 15, 0)` check, building Prism Launcher against Qt 5.14 would raise a deprecation warning, which would then abort the build as you have added `-Werror` (and MSVC equivalents)

`-Werror` and msvc equivalent `/WX` are locked behind the setting `WARNINGS_AS_ERRORS`
who ever is building would have to explicitly set `WARNINGS_AS_ERRORS=yes` for them to be added.

**Comment by Scrumplex** [comment]
> `-Werror` and msvc equivalent `/WX` are locked behind the setting `WARNINGS_AS_ERRORS`
> who ever is building would have to explicitly set `WARNINGS_AS_ERRORS=yes` for them to be added.

Most packagers would define these kinds of flags using `CMAKE_CXX_FLAGS`. Many build tools (Nix, Makepkg, Flatpak-Builder) set these variables (sometimes indirectly) anyway, so I don't think we *need* this if we don't plan to make it the default.

**Comment by Ryex** [comment]
hoo

I just cleaned up all the warnings I get from MSVC

I think we might want to add [ac89ac3](https://github.com/PrismLauncher/PrismLauncher/pull/1107/commits/ac89ac37f03571fe3f9e0639f1a7377463b60870)  and [cc41b03](https://github.com/PrismLauncher/PrismLauncher/pull/1107/commits/cc41b039e6ba2a24c0ded9fb573bee6050aa2ac9) to our blame ignore
when this is merged but this really is ready I think.

I've removed the WARNINGS_AS_ERRORS option

File: launcher/minecraft/PackProfile.cpp [hunk 1]
```diff
@@ -532,9 +532,9 @@
     return (*iter);
 }
 
-ComponentPtr PackProfile::getComponent(int index)
+ComponentPtr PackProfile::getComponent(size_t index)
 {
-    if(index < 0 || index >= d->components.size())
+    if(index < 0 || index >= static_cast<size_t>(d->components.size()))
```
[the remark this answers is not in the document]
Comment (Scrumplex) [reply launcher/minecraft/PackProfile.cpp]: hmm size_t is unsigned I think?

**Review by Scrumplex — APPROVED** [review]

File: launcher/minecraft/PackProfile.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (Ryex) [reply launcher/minecraft/PackProfile.cpp]: the explicit cast was to suppress a warning about data loss in comparing different types.
so yes, if `size_t` unsigned `index < 0` would always be false.

[the remark this answers is not in the document]
Comment (Scrumplex) [reply launcher/minecraft/PackProfile.cpp]: Yes. But you changed the function signature to use size_t so the `index < 0` doesn't do anything

File: launcher/minecraft/PackProfile.cpp [hunk 2]
```diff
@@ -533,9 +533,9 @@ ComponentPtr PackProfile::getComponent(const QString &id)
...
```
Comment (Ryex) [inline launcher/minecraft/PackProfile.cpp]:
```suggestion
    if(index >= static_cast<size_t>(d->components.size()))
```

**Review by Scrumplex — APPROVED** [review]

**Comment by Scrumplex** [comment]
Oh. Clangd added a bad include there oops.

File: CMakeLists.txt [hunk 3]
```diff
@@ -85,38 +92,40 @@ set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -DTOML_ENABLE_FLOAT16=0")
 # set CXXFLAGS for build targets
 set(CMAKE_CXX_FLAGS_RELEASE "-O2 -D_FORTIFY_SOURCE=2 ${CMAKE_CXX_FLAGS_RELEASE}")
 
+
 option(DEBUG_ADDRESS_SANITIZER "Enable Address Sanitizer in Debug builds" on)
 
 # If this is a Debug build turn on address sanitiser
-if (CMAKE_BUILD_TYPE STREQUAL "Debug" AND DEBUG_ADDRESS_SANITIZER)
+if ((CMAKE_BUILD_TYPE STREQUAL "Debug" OR CMAKE_BUILD_TYPE STREQUAL "RelWithDebInfo") AND DEBUG_ADDRESS_SANITIZER)
```
Comment (getchoo) [inline CMakeLists.txt:98]: everything else looks good, but i'm really not sure about this...address sanitizer is great, but does have some overhead associated with it; and it's not exactly uncommon for `RelWithDebInfo` to be used by some linux distributions or even users. imo, a better option would be to use asan by default on debug builds, while keeping it as an option for `RelWithDebInfo` or even regular `Release` builds

**Review by getchoo — CHANGES_REQUESTED** [review]

File: CMakeLists.txt (hunk 3 above)
Comment (Trial97) [reply CMakeLists.txt:98]: I do believe that address sanitizer should not be on by default for any kind of build, but instead, explicitly activate them where needed.

**Review by Trial97 — APPROVED** [review]

**Review by txtsd — APPROVED** [review]

File: CMakeLists.txt (hunk 3 above)
Comment (getchoo) [reply CMakeLists.txt:98]: yeah this would make sense after https://github.com/PrismLauncher/PrismLauncher/pull/1518

Comment (txtsd) [reply CMakeLists.txt:98]: @getchoo could you hit Resolve so we can merge?

**Review by getchoo — APPROVED** [review]

**Pull request merged**